packages/queued-request-controller/src/QueuedRequestMiddleware.ts

```diff
@@ -0,0 +1,140 @@
+  NetworkControllerGetStateAction,
+  NetworkControllerSetActiveNetworkAction,
+  NetworkControllerSetProviderTypeAction,
+} from '@metamask/network-controller';
+import type { SelectedNetworkControllerSetNetworkClientIdForDomainAction } from '@metamask/selected-network-controller';
+import { SelectedNetworkControllerActionTypes } from '@metamask/selected-network-controller';
+
+import type { QueuedRequestControllerEnqueueRequestAction } from './QueuedRequestController';
+import { QueuedRequestControllerActionTypes } from './QueuedRequestController';
+
+const isConfirmationMethod = (method: string) => {
+  const confirmationMethods = [
+    'eth_sendTransaction',
+    'wallet_watchAsset',
+    'wallet_switchEthereumChain',
+    'eth_signTypedData_v4',
+  ];
+
+  return !confirmationMethods.includes(method);
+};
+
+export const createQueuedRequestMiddleware = (
+  messenger: ControllerMessenger<
+    | QueuedRequestControllerEnqueueRequestAction
+    | NetworkControllerGetStateAction
+    | NetworkControllerSetActiveNetworkAction
+    | NetworkControllerSetProviderTypeAction
+    | NetworkControllerGetNetworkClientByIdAction
+    | NetworkControllerFindNetworkClientIdByChainIdAction
+    | SelectedNetworkControllerSetNetworkClientIdForDomainAction
+    | AddApprovalRequest,
+    never
+  >,
+  useRequestQueue: () => boolean,
+): JsonRpcMiddleware<any, any> => {
+  return createAsyncMiddleware(async (req: any, res: any, next) => {
+    if (!useRequestQueue() || isConfirmationMethod(req.method)) {
+      next();
+      return;
+    }
+
+    const networkClientIdForRequest = req.networkClientId as NetworkClientId;
+
+    await messenger.call(
+      QueuedRequestControllerActionTypes.enqueueRequest,
+      async () => {
+        if (req.method === 'wallet_switchEthereumChain') {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        const isBuiltIn = isNetworkType(networkClientIdForRequest);
+        let networkConfigurationForRequest;
+        if (isBuiltIn) {
+          const builtIn = BUILT_IN_NETWORKS[networkClientIdForRequest];
+          if (builtIn.chainId) {
+            networkConfigurationForRequest = builtIn;
+          }
+        }
+
+        networkConfigurationForRequest =
+          networkConfigurationForRequest ||
+          messenger.call(
+            'NetworkController:getNetworkClientById',
+            networkClientIdForRequest,
+          ).configuration;
+
+        const currentProviderConfig = messenger.call(
+          'NetworkController:getState',
+        ).providerConfig;
+
+        const currentChainId = currentProviderConfig.chainId;
+
+        if (currentChainId === networkConfigurationForRequest.chainId) {
+          // eslint-disable-next-line n/callback-return
+          return next();
+        }
+
+        // if is switch eth chain call
+        // clear request queue when the switch ethereum chain call completes (success or fail)
```

Comment:
even in a cancelled switch call? Even if the request is coming from a domain which wasn't expecting a switch? Shouldn't we see that request still since we're still on the right chain?

Reply:
was this resolved?

Reply:
yup! https://github.com/MetaMask/core/pull/1806/files#diff-02c32f437b122b5c71f0ac6bdf304cd22f6d4569c9c9b8ecbb711c11c98cd0d3R120

Reply:
(at least I think the comment addresses your concern. The main take-away though is that what is decribed in this comment is not current implmented.)